src/pure_type.erl

```diff
@@ -19,11 +19,40 @@
 %%
 %% -------------------------------------------------------------------
 
+%% @doc Inspired by both the Lasp data type behaviour and the Riak data
+%%      type behaviour.
+
 -module(pure_type).
--author("Vitor Enes Duarte <vitorenesduarte@gmail.com>").
+-author("Georges Younes <georges.r.younes@gmail.com>").
+
+%-export([update/3]).
+-export_type([polog/0, version_vector/0, element/0]).
 
--export_type([pure_type/0]).
 
 %% Define some initial types.
--type pure_type() ::  pure_gcounter |
-                      pure_gset.
+-type pure_type() :: pure_gcounter | pure_pncounter | pure_gset | pure_twopset | pure_aworset.
+-type polog() :: orddict:orddict().
+-type pure_payload() :: {polog(), term()}.
+-type version_vector() :: orddict:orddict().
+-type pure_crdt() :: {pure_type(), pure_payload()}.
+-type pure_operation() :: term().
+-type element() :: term().
+-type value() :: term().
+-type error() :: term().
+
+%% Initialize a Pure op-based CRDT.
+-callback new() -> pure_crdt().
+
+%% Unified interface for allowing parameterized Pure op-based CRDTs.
+-callback new([term()]) -> pure_crdt().
+
+%% Perform an update.
+-callback update(pure_operation(), version_vector(), pure_crdt()) ->
```

Comment:
yes, I will probably do the changes tonight or tomorrow.

Thanks,
Georges
On 24 May 2016 15:58, "Christopher S. Meiklejohn" <notifications@github.com>
wrote:

> In src/pure_type.erl
> <https://github.com/lasp-lang/types/pull/31#discussion_r64407780>:
>
> > +-type pure_payload() :: {polog(), term()}.
> > +-type version_vector() :: orddict:orddict().
> > +-type pure_crdt() :: {pure_type(), pure_payload()}.
> > +-type pure_operation() :: term().
> > +-type element() :: term().
> > +-type value() :: term().
> > +-type error() :: term().
> > +
> > +%% Initialize a Pure op-based CRDT.
> > +-callback new() -> pure_crdt().
> > +
> > +%% Unified interface for allowing parameterized Pure op-based CRDTs.
> > +-callback new([term()]) -> pure_crdt().
> > +
> > +%% Perform an update.
> > +-callback update(pure_operation(), version_vector(), pure_crdt()) ->
>
> Should this be called mutate to match with the existing API?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lasp-lang/types/pull/31/files/4d2a439ec3d8789bf8424861b8c9f4caa33e0d15#r64407780>
>